spark/firmware
# v2 dcd format and tests

## Do not merge as-is, will require a rebase

### Problem

Hypothesised write errors in flash that appear only after power cycle can corrupt the configuration data.
Flash access is not thread-safe.

### Solution

Adds CRC checking to the Electron DCD implementation so that write errors are detected.
Add a critical section around flash operations and around DCD operations.

### Steps to Test

Run the firmware unit tests.

Manual testing:

- flash system firmware to an electron that was connecting to the cloud
- verify that the device continues to connect to the cloud
- make a change in DCD via DFU
- verify the device still connects to the cloud (the DCD contents will have been upgraded to the V2 format.)

### Example App

This is an internal system change. No example application.

---

### Completeness
